src/test/GMRESSolverTests.cpp

```diff
@@ -0,0 +1,110 @@
+#include "GMRESSolver.hpp"
+#include <igloo/igloo_alt.h>
+#include <iostream>
+
+using namespace igloo;
+using namespace blitzdg;
+
+using std::cout;
+
+namespace GMRESSolverTests {
+    class Matvec {
+        const matrix_type& mref_;
+    public:
+        Matvec(const matrix_type& mat)
+            : mref_{ mat }
+        {}
+
+        bool operator()(const vector_type& in, vector_type& out) const {
+            blitz::firstIndex i;
+            blitz::secondIndex j;
+            out = blitz::sum(mref_(i, j) * in(j), j);
+            return true;
+        }
+    };
+
+    // identity preconditioner
+    class Precon {
+    public:
+        bool operator()(const vector_type& in, vector_type& out) const {
+            out = in;
+            return true;
+        }
+    };
+
+    GMRESSolver* solver = nullptr;
+
+    Describe(GMRESSolver_Object) {
+        void Setup() {
+            solver = new GMRESSolver();
+        }
+
+        It(Solve_Ax_equals_b) {
+            index_type N = 5;
+            vector_type b(N), x(N);
+            matrix_type A(N, N);
+
+            A = 2.,3.,0.,0.,0.,
+                3.,0.,4.,0.,6.,
+                0.,-1.,-3.,2.,0.,
+                0.,0.,1.,0.,0.,
+                0.,4.,2.,0.,1.;
+
+            b =  8.,
+                45.,
+                -3.,
+                3.,
+                19.;
+
+            x = 1.,
+                2.,
+                3.,
+                4.,
+                5.;
+
+            cout << "GMRESSolver: nonsingular system\n";
+            GMRESSolver& gmres = *solver;
+            vector_type soln(N);
+            for (index_type i = 0; i < N; ++i)
+                soln(i) = rand() / (real_type(1) + RAND_MAX);
```

Comment:
Why the random numbers? Is this for initial guess?

Reply:
Yah, just wanted to test it with a nonzero initial guess.

Reply:
I believe that blitz uses reference semantics for copy construction. So, the member variable Matvec::mat_ should be created as an alias of the input matrix. Because of this behavior, I don't think we need to worry about any overhead incurred by passing the Matvec functor by value. Clearly, this isn't an issue here, since A is only 5 x 5, but it would be good to know in general. There is also the reference() function in blitz, which may do the same thing.